gettime.c

```diff
@@ -127,10 +128,58 @@ static void fio_init gtod_init(void)
 
 #endif /* FIO_DEBUG_TIME */
 
+#if defined(CONFIG_CLOCK_MONOTONIC_RAW) && defined(CONFIG_CLOCK_MONOTONIC)
+static uint64_t bench_clock_gettime(clockid_t clkid, int rounds)
+{
+	int i;
+	struct timespec start, ts;
+
+	fio_gettime(&start, NULL);
+	for (i = 0; i < rounds; ++i)
+		clock_gettime(clkid, &ts);
+	return utime_since_now(&start);
+}
+
+/*
+ * There are platforms using Linux kernels prior to 5.3 where
+ * clock_gettime(CLOCK_MONOTONIC, ...) is implemented via vDSO but
+ * clock_gettime(CLOCK_MONOTONIC_RAW, ...) is not (and falls back to a "slow"
+ * syscall) so check for a dramatic difference in reported times.
+ */
+static bool cg_monotonic_raw_slow(void)
+{
+	uint64_t bench_monotonic_raw, bench_monotonic;
+	const int rounds = 1000;
+
+	bench_monotonic = bench_clock_gettime(CLOCK_MONOTONIC, rounds);
+	bench_monotonic_raw = bench_clock_gettime(CLOCK_MONOTONIC_RAW, rounds);
+	dprint(FD_TIME,
+	       "CLOCK_MONOTONIC=%" PRIu64 "us, CLOCK_MONOTONIC_RAW=%" PRIu64
+	       "us (%d rounds)\n",
+	       bench_monotonic, bench_monotonic_raw, rounds);
+	if ((bench_monotonic * 2) < bench_monotonic_raw) {
+		dprint(FD_TIME,
+		       "Forcing CLOCK_MONOTONIC as clock_gettime()'s clock\n");
+		return true;
+	} else
+		return false;
+}
+#else
+static bool cg_monotonic_raw_slow(void)
+{
+	return false;
+}
+#endif /* defined(CONFIG_CLOCK_MONOTONIC_RAW) && defined(CONFIG_CLOCK_MONOTONIC) */
+
 #ifdef CONFIG_CLOCK_GETTIME
 static int fill_clock_gettime(struct timespec *ts)
 {
-#if defined(CONFIG_CLOCK_MONOTONIC_RAW)
+#if defined(CONFIG_CLOCK_MONOTONIC_RAW) && defined(CONFIG_CLOCK_MONOTONIC_RAW)
+	if (cg_prefer_monotonic)
```

Comment:
I'm not convinced we need to do this, but if we do, there's a trivial way to avoid this branch - just have the tester set the preferred clocksource, and don't do the check here.